### ISSUE TITLE
Volatile functions in subqueries in UPDATE incorrectly evaluated on coordinator

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#2148]
If I run the following command on the master branch:

```
CREATE TABLE sessions(session_id int, priority double precision);
SELECT create_distributed_table('sessions', 'session_id');
INSERT INTO sessions VALUES (1);
INSERT INTO sessions VALUES (2);
UPDATE sessions s SET priority = r.random FROM (SELECT session_id, random() FROM sessions) r WHERE s.session_id = r.session_id;

SELECT * FROM sessions;
 session_id │     priority      
────────────┼───────────────────
          1 │ 0.706227001734078
          2 │ 0.706227001734078
(2 rows)
```

All rows are assigned the same priority, but the Postgres behaviour is to re-evaluate `random()` for every row. The problem is that the coordinator evaluates the `random()`, though we should be erroring out when there's a volatile function.

**Comment by metdos** [comment]
@marcocitus, I suppose we want to backport this to 7.4?

**Comment by marcocitus** [comment]
Yep

**Comment by marcocitus** [comment]
Fixed by #2150

**Comment by marcocitus** [comment]
It seems this bug still exists in a different form:
```sql
UPDATE test SET y = 3 WHERE x = 3 AND y IN (SELECT y FROM test WHERE x = 3 AND y > random());
LOG:  issuing UPDATE public.test_102738 test SET y = 3 WHERE ((x OPERATOR(pg_catalog.=) 3) AND (y OPERATOR(pg_catalog.=) ANY (SELECT test_1.y FROM public.test_102738 test_1 WHERE ((test_1.x OPERATOR(pg_catalog.=) 3) AND ((test_1.y)::double precision OPERATOR(pg_catalog.>) '0.731105228886008'::double precision)))))
DETAIL:  on server localhost:9701
```
